upgrade.sh

```diff
@@ -52,6 +52,66 @@ function unbound_psp_check() {
     echo "cray-unbound-coredns-psp check Done"
 }
 
+# Function to check cray-sysmgmt-health chart with app version 9.3.1 for prometheus-operator and retain old PVs data.
+function sysmgmt_health() {
+    echo "Checking for chart version of cray-sysmgmt-health"
+    version="45.1"
+    if [ $(helm ls -o json --namespace sysmgmt-health| jq -r --argjson version $version '.[] | select(.app_version | sub(".[0-9]$";"") | tonumber | . < $version).name') ]
+    then
+    prom_pvc="prometheus-cray-sysmgmt-health-promet-prometheus-db-prometheus-cray-sysmgmt-health-promet-prometheus-0"
+    alert_pvc="alertmanager-cray-sysmgmt-health-promet-alertmanager-db-alertmanager-cray-sysmgmt-health-promet-alertmanager-0"
+    echo "Get PV for both prometheus and Alertmanager"
+    prom_pv=$(kubectl get pvc -n sysmgmt-health --output=json $prom_pvc |jq '.spec."volumeName"')
+    alert_pv=$(kubectl get pvc -n sysmgmt-health --output=json $alert_pvc |jq '.spec."volumeName"')
+    prom_pv="${prom_pv//[\",]}"
+    alert_pv="${alert_pv//[\",]}"
+    echo "$prom_pv"
+    echo "$alert_pv"
+
+    # Patch the PersistenceVolume created/used by the prometheus-operator and alertmanager to Retain claim policy
+    kubectl patch pv/$prom_pv -p '{"spec":{"persistentVolumeReclaimPolicy":"Retain"}}'
+    kubectl patch pv/$alert_pv -p '{"spec":{"persistentVolumeReclaimPolicy":"Retain"}}'
+
+    # Uninstall the cray-sysmgmt-health release
+    helm ls -o json --namespace sysmgmt-health| jq -r --argjson version $version '.[] | select(.app_version | sub(".[0-9]$";"") | tonumber | . < $version).name'| xargs -L1 helm uninstall --namespace sysmgmt-health
+    
+    # Delete the existing PersistentVolumeClaim, and verify PV become Released
+    kubectl delete pvc/$prom_pvc -n sysmgmt-health
+    kubectl delete pvc/$alert_pvc -n sysmgmt-health
+    alert_pv_phase=$(kubectl get pv $prom_pv -o json |jq '.status."phase"')
+    prom_pv_phase=$(kubectl get pv $alert_pv -o json |jq '.status."phase"')
+    prom_pv_phase="${prom_pv_phase//[\",]}"
+    alert_pv_phase="${alert_pv_phase//[\",]}"
+    echo "Verifying whether PVs became Released or not."
+    if [ "$alert_pv_phase" == Released ] && [ "$prom_pv_phase" == Released ]; then
+        echo "Both Prometheus and Alertmanager PVs are Released"
+    else
+        echo >&2 "PVs are not Released. Verify if PV exists or not."
+    fi
+    # Remove the cray-sysmgmt-health-promet-kubelet service.
+    echo "Deleting cray-sysmgmt-health-promet-kubelet service in kube-system namespace."
+    kubectl delete service/cray-sysmgmt-health-promet-kubelet -n kube-system
+
+    echo "Deleting sysmgmt-health existing CRDs"
+    # Remove all the existing CRDs (ServiceMonitors, Podmonitors, etc.)
+    kubectl delete crd alertmanagers.monitoring.coreos.com podmonitors.monitoring.coreos.com prometheuses.monitoring.coreos.com prometheusrules.monitoring.coreos.com servicemonitors.monitoring.coreos.com thanosrulers.monitoring.coreos.com
```

Comment:
Could probably make this a bit less hard coded and search for the crds owning helm chart. Something like:

```
  for h in helm-chart-name; do
    for c in $(kubectl get crds -A -o jsonpath='{range .items[?(@.metadata.annotations.meta\.helm\.sh\/release\-name=="'${h}'")]}{.metadata.name}{"\n"}{end}'); do
      kubectl delete crd ${c}
    done
  done
```

Reply:
In sysmgmt-health namespace or cray-sysmgmt-health chart there are total 101 CRDs and we need to delete just 6 CRDs out of them, so we can't use logic for deleting CRDs from particular namespace or chart.
Using regex also won't help here as there are CRDs with similar names.
So we hardcoded all 6 CRDs that needs to be deleted.

Reply:
Crd's installed via helm are owned by a specific namespace/name of the helm chart. That isn't stating they are owned by a namespace. Are you saying these crd's are not installed via helm? If not why not? What controls the CRD installation as helm can handle that directly. I'd say we should be managing crd's via the helm chart directly and not hacking in crd removal and updates in scripts.

Example:
```
ncn-m001:~ # for h in cert-manager-init; do     for c in $(kubectl get crds -A -o jsonpath='{range .items[?(@.metadata.annotations.meta\.helm\.sh\/release\-namespace=="'${h}'")]}{.metadata.name}{"\n"}{end}'); do       echo helm owned crd ${c};     done; done 
helm owned crd certificaterequests.cert-manager.io
helm owned crd certificates.cert-manager.io
helm owned crd challenges.acme.cert-manager.io
helm owned crd clusterissuers.cert-manager.io
helm owned crd issuers.cert-manager.io
helm owned crd orders.acme.cert-manager.io
```

Reply:
Looking at things that indeed is the case, why are these crd's not owned by a chart? We can have helm manage the upgrade of the crd resources directly if we do that.

```
ncn-m001:~ # kubectl get crd orders.acme.cert-manager.io -o=jsonpath='{.metadata.annotations}' | jq
{
  "cert-manager.io/inject-ca-from-secret": "cert-manager/cray-certmanager-cert-manager-webhook-tls",
  "meta.helm.sh/release-name": "cray-certmanager-init",
  "meta.helm.sh/release-namespace": "cert-manager-init"
}
ncn-m001:~ # kubectl get crd alertmanagers.monitoring.coreos.com -o=jsonpath='{.metadata.annotations}' | jq
{
  "controller-gen.kubebuilder.io/version": "v0.2.4"
}
```

Reply:
Note we can have helm "take ownership" of updates to resources via the same logic in this pr:
https://github.com/helm/helm/pull/7649

Reply:
For cert-manager CRDs you are using annotation (metadata.annotations.meta\.helm\.sh\/release\-namespace=="'${h}'")

Name:         certificaterequests.cert-manager.io
Namespace:
Labels:       app.kubernetes.io/managed-by=Helm
Annotations:  cert-manager.io/inject-ca-from-secret: cert-manager/cray-certmanager-cert-manager-webhook-tls
              meta.helm.sh/release-name: cray-certmanager-init
              meta.helm.sh/release-namespace: cert-manager-init

But for all CRDs these annotations are not configured.

But we found an annotation which is common in all the CRDs that needs to be deleted .

`for c in $(kubectl get crds -A -o jsonpath='{range .items[?(@.metadata.annotations.controller-gen\.kubebuilder\.io\/version=="v0.2.4")]}{.metadata.name}{"\n"}{end}'); do
      kubectl delete crd ${c}
    done

Reply:
Correct, it is helm that actually is adding the annotations. But by adding them then helm can manage updates to them and we don't have to kubectl delete crd and then add.